File: implementations/python/sdk/tests/validate.py
from __future__ import annotations

import json
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
SDK_ROOT = HERE.parent
SRC_ROOT = SDK_ROOT / "src"
MANIFEST = SDK_ROOT.parent.parent / "sdk" / "protocol" / "ai-sdk-protocol-v1.json"
sys.path.insert(0, str(SRC_ROOT))

from multiplexed_ai_sdk import (  # noqa: E402
    AI_SDK_OPERATION_RETRY,
    AI_SDK_OPERATIONS,
    AI_SDK_PROTOCOL_VERSION,
    AiSdkExecutionMode,
    AiSdkExecutionStatus,
    AiSdkInvocationKind,
    AiSdkPublicationDependencyPackageKind,
    AiSdkPublicationFunctionKind,
)

manifest = json.loads(MANIFEST.read_text(encoding="utf-8"))
expected_operations = [item["name"] for item in manifest["operations"]]
assert AI_SDK_PROTOCOL_VERSION == manifest["protocolVersion"]
assert list(AI_SDK_OPERATIONS.values()) == expected_operations
assert AI_SDK_OPERATION_RETRY["sdk.execution.observe"] == "safe-read"
assert AI_SDK_OPERATION_RETRY["sdk.execution.result"] == "safe-read"
for name in ("sdk.publish_pipeline", "sdk.execution.submit", "sdk.execution.cancel"):
    assert AI_SDK_OPERATION_RETRY[name] == "never"

assert [item.value for item in AiSdkExecutionMode] == ["Sequential", "Dag"]
assert [item.value for item in AiSdkInvocationKind] == ["Native", "Custom", "Mcp"]
assert [item.value for item in AiSdkExecutionStatus] == [
    "Pending", "Running", "Waiting", "Completed", "Failed", "Cancelled"
]
assert [item.value for item in AiSdkPublicationFunctionKind] == [
    "Step", "ConcurrencyPolicy", "RetryPolicy", "DelegationPolicy"
]
assert [item.value for item in AiSdkPublicationDependencyPackageKind] == [
    "PythonWheelBundle", "NodeLockedBundle", "DotNetAssemblyClosure"
]

for forbidden in (
    "Multiplexed.AI",
    "MongoDB",
    "StackExchange.Redis",
    "TenantId",
    "TenantGroupId",
    "SharedRunId",
    "LocalRunId",
    "RuntimeInstanceId",
    "WorkerId",
    "ClaimToken",
    "AssignmentEpoch",
    "ControlPlaneId",
):
    for path in SRC_ROOT.rglob("*.py"):
        assert forbidden not in path.read_text(encoding="utf-8"), f"Forbidden token {forbidden} in {path}"

print(f"Python external SDK validated: {len(expected_operations)} operations.")

File: implementations/node/sdk/tests/validate.mjs
import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";
import ts from "../../workers/hosted_invocation/vendor/typescript-5.8.3.cjs";

const here = path.dirname(fileURLToPath(import.meta.url));
const sdkRoot = path.resolve(here, "..");
const sourceRoot = path.join(sdkRoot, "src");
const manifestPath = path.resolve(sdkRoot, "../../sdk/protocol/ai-sdk-protocol-v1.json");

const sourceFiles = [];
const collect = (directory) => {
  for (const entry of fs.readdirSync(directory, { withFileTypes: true })) {
    const fullPath = path.join(directory, entry.name);
    if (entry.isDirectory()) collect(fullPath);
    else if (entry.isFile() && entry.name.endsWith(".ts")) sourceFiles.push(fullPath);
  }
};
collect(sourceRoot);

for (const file of sourceFiles) {
  const source = fs.readFileSync(file, "utf8");
  const result = ts.transpileModule(source, {
    fileName: file,
    reportDiagnostics: true,
    compilerOptions: {
      target: ts.ScriptTarget.ES2022,
      module: ts.ModuleKind.ESNext,
      strict: true,
    },
  });
  const errors = (result.diagnostics ?? []).filter(
    (diagnostic) => diagnostic.category === ts.DiagnosticCategory.Error,
  );
  if (errors.length > 0) {
    for (const diagnostic of errors) {
      console.error(ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n"));
    }
    process.exit(1);
  }
}

const manifest = JSON.parse(fs.readFileSync(manifestPath, "utf8"));
const operationSource = fs.readFileSync(
  path.join(sourceRoot, "protocol/operations.ts"),
  "utf8",
);
for (const operation of manifest.operations) {
  if (!operationSource.includes(`"${operation.name}"`)) {
    throw new Error(`TypeScript protocol is missing operation '${operation.name}'.`);
  }
}
if (!operationSource.includes(`AI_SDK_PROTOCOL_VERSION = ${manifest.protocolVersion}`)) {
  throw new Error("TypeScript protocol version does not match the canonical manifest.");
}

const packageJson = JSON.parse(fs.readFileSync(path.join(sdkRoot, "package.json"), "utf8"));
if (packageJson.dependencies?.["@modelcontextprotocol/client"] !== "2.0.0") {
  throw new Error("TypeScript SDK must pin the external MCP client dependency.");
}
if (packageJson.engines?.node !== ">=20") {
  throw new Error("TypeScript SDK Node requirement must remain capability-based at >=20.");
}

const forbidden = [
  "Multiplexed.Abstractions",
  "MongoDB",
  "StackExchange.Redis",
  "TenantId",
  "TenantGroupId",
  "SharedRunId",
  "LocalRunId",
  "RuntimeInstanceId",
  "WorkerId",
  "ClaimToken",
  "AssignmentEpoch",
  "ControlPlaneId",
];
for (const file of sourceFiles) {
  const source = fs.readFileSync(file, "utf8");
  for (const token of forbidden) {
    if (source.includes(token)) {
      throw new Error(`Forbidden runtime/internal token '${token}' found in ${file}.`);
    }
  }
}

console.log(
  `TypeScript SDK validated: ${sourceFiles.length} source files, ${manifest.operations.length} operations.`,
);
